java-checks/src/main/java/org/sonar/java/checks/ObjectFinalizeOverridenCallsSuperFinalizeCheck.java

```diff
@@ -113,13 +116,34 @@ private static boolean isLastStatement(@Nullable BlockTree blockTree, MethodInvo
     return false;
   }
 
-  private static boolean isFinalize(MethodTree methodTree) {
-    if ("finalize".equals(methodTree.simpleName().name()) && methodTree.parameters().isEmpty()) {
-      Tree returnType = methodTree.returnType();
-      if (returnType != null && returnType.is(Tree.Kind.PRIMITIVE_TYPE)) {
-        return "void".equals(((PrimitiveTypeTree) returnType).keyword().text());
+  private static boolean isFinalize(Symbol symbol) {
+    if (FINALIZE.equals(symbol.name()) && symbol.isMethodSymbol()) {
+      Symbol.MethodSymbol methodSymbol = (Symbol.MethodSymbol) symbol;
+      if ("void".equals(methodSymbol.returnType().name()) && methodSymbol.parameterTypes().isEmpty()) {
+        return true;
       }
     }
     return false;
   }
+
+  private static boolean doesOverrideFinalize(Symbol classSymbol) {
+    if (classSymbol.isTypeSymbol()) {
+      Type superClassType = ((Symbol.TypeSymbol) classSymbol).superClass();
+      if (superClassType != null) {
+        Symbol.TypeSymbol currentClass = superClassType.symbol();
+        superClassType = currentClass.superClass();
+        while (superClassType != null) {
+          for (Symbol symbol : currentClass.lookupSymbols(FINALIZE)) {
+            if (isFinalize(symbol)) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)